### ISSUE TITLE
README.md

### ISSUE THREAD

**Issue by Ganesh-Sai24** [Ganesh-Sai24/explainable-ensemble-fetal-health#1]
# An Optimized Explainable Ensemble Learning Framework for AI-Assisted Early Fetal Health Risk Assessment Using Cardiotocography Data

## 📌 Project Overview

This project proposes an **optimized and explainable ensemble learning framework** for the early assessment of fetal health risk using **Cardiotocography (CTG) data**.

Cardiotocography is a commonly used fetal monitoring technique that records important physiological information, primarily:

- Fetal Heart Rate (FHR)
- Uterine Contractions (UC)

CTG data can provide valuable information about fetal condition. However, manual interpretation of CTG recordings can be complex and may be affected by inter-observer variability and the large amount of information that needs to be analyzed.

Machine Learning (ML) provides an opportunity to learn patterns from CTG-derived features and assist in classifying fetal conditions.

The proposed project builds upon previous CTG-based machine learning research by focusing on:

- Class imbalance handling
- Ensemble learning
- Model optimization
- Explainable Artificial Intelligence (XAI)
- Risk-oriented prediction
- Clinical decision-support presentation

---

## 🎯 Project Objective

The primary objective of this project is to develop an **optimized, explainable ensemble learning framework** that can analyze CTG-derived features and assist in the early assessment of fetal health risk.

The framework aims to:

1. Preprocess and analyze CTG data.
2. Handle class imbalance between fetal health categories.
3. Identify relevant and informative features.
4. Train multiple machine learning models.
5. Combine complementary models using ensemble learning.
6. Optimize the selected models and their parameters.
7. Generate fetal health risk predictions.
8. Provide explanations for model predictions using XAI techniques.
9. Present the prediction and explanation through a user-friendly interface.
10. Support clinical decision-making without replacing healthcare professionals.

---

## 🏥 Problem Statement

Fetal health monitoring is an important part of prenatal care.

Cardiotocography provides information about fetal heart rate and uterine contractions, which can be used to assess fetal condition.

However, CTG datasets can contain **imbalanced classes**, where normal fetal records are much more numerous than suspect or pathological records.

For example, the UCI Cardiotocography dataset used in the base research contains:

- **2,126 records**
- **21 input attributes**
- **3 fetal health classes**
  - Normal
  - Suspect
  - Pathological

The class distribution is highly uneven, with the Normal class representing the majority of records.

If a model is evaluated only using conventional accuracy, it may appear to perform well while giving inadequate attention to minority classes.

Therefore, a fetal-health prediction system should consider:

- Class imbalance
- Minority-class detection
- Balanced evaluation metrics
- Model generalization
- Interpretability
- Clinical usability

---

## 💡 Motivation

The project is motivated by three major observations from existing research.

### 1. CTG-based machine learning

Previous research has demonstrated that machine learning algorithms can classify fetal health conditions from CTG-derived features.

The base paper investigated multiple machine learning algorithms and a deep neural network while addressing class imbalance using SMOTE and evaluating performance using balanced accuracy.

### 2. Broader applications of AI in fetal assessment

Recent research shows that AI is being applied to fetal assessment using multiple data modalities, including:

- Ultrasound
- Doppler measurements
- Maternal characteristics
- Biochemical markers
- Genetic information

These studies demonstrate the potential of AI for early fetal risk assessment while also highlighting challenges such as external validation, generalization, bias and clinical translation.

### 3. Explainable and clinically integrated AI

Recent reviews of AI in women's health emphasize that clinical AI should not only produce predictions but should also consider:

- Explainability
- Transparency
- Bias
- Privacy
- Regulation
- Workflow integration
- Clinical decision support

Therefore, this project focuses on combining predictive performance with interpretability and practical usability.

---

## 📚 Literature Foundation

This project is primarily informed by the following research directions.

### Base Paper

**Early Detection of Fetal Health Conditions Using Machine Learning for Classifying Imbalanced Cardiotocographic Data**

Nazli et al., *Diagnostics*, 2025.

The study investigated fetal health classification using CTG data and compared several machine learning algorithms and a deep neural network.

The study focused strongly on the problem of class imbalance and used **SMOTE (Synthetic Minority Oversampling Technique)** to improve representation of minority classes.

The study also emphasized **Balanced Accuracy** rather than relying only on conventional accuracy.

Among the evaluated models, LightGBM achieved the highest reported 5-fold cross-validation balanced accuracy of **91.34%**.

The study provides the primary technical foundation for this project.

---

### Supporting Paper 1

**Application of artificial intelligence in diagnosis and management of fetal growth disorders: a comprehensive review**

Ługowski, Babińska and Stanirowski, *Frontiers in Medicine*, 2026.

This review discusses the application of AI and ML in:

- Fetal Growth Restriction (FGR)
- Small-for-Gestational-Age (SGA)
- Large-for-Gestational-Age (LGA)
- Fetal macrosomia

The review describes AI applications involving ultrasound, Doppler, maternal characteristics, biochemical markers and genomic information.

It also highlights important challenges such as:

- Retrospective datasets
- Limited sample sizes
- Lack of external validation
- Generalizability
- Bias
- Infrastructure
- Data privacy
- Regulatory requirements

This paper provides the broader clinical and research context for AI-based fetal assessment.

---

### Supporting Paper 2

**AI-Driven Advances in Women's Health Diagnostics: Current Applications and Future Directions**

Christian Macedonia, *Diagnostics*, 2025.

This narrative review discusses AI applications across obstetrics and gynecology, including:

- AI-enhanced fetal ultrasound
- Cervical screening
- Preeclampsia prediction
- Endometriosis diagnosis
- Remote maternal-fetal monitoring
- Digital biomarkers
- Clinical decision support

The paper emphasizes the importance of:

- Explainability
- Transparency
- Bias mitigation
- Privacy
- Regulatory considerations
- Workflow integration
- Equity

It also emphasizes the principle that AI should **augment clinical judgment rather than replace clinicians**.

This paper provides the conceptual foundation for incorporating explainability and clinical decision support into the proposed framework.

---

## 🔬 Research Gap

Based on the reviewed literature, several research challenges can be identified.

### Gap 1 — Class imbalance

Fetal health datasets may contain significantly fewer pathological and suspect cases compared with normal cases.

This can make conventional accuracy insufficient for evaluating model performance.

### Gap 2 — Dependence on individual models

Different machine learning algorithms can capture different patterns in CTG data.

Relying on one classifier may not fully exploit the complementary strengths of different algorithms.

### Gap 3 — Limited interpretability

A prediction such as:

    Pathological — 87%

does not by itself explain why the model produced that prediction.

For clinical decision support, understanding the contributing features can be important.

### Gap 4 — Optimization

Machine learning performance depends on model configuration, preprocessing, feature representation and hyperparameters.

Systematic optimization can therefore be investigated as part of the proposed framework.

### Gap 5 — Clinical usability

A prediction model should ideally provide information in a form that can be understood and reviewed by users.

Therefore, the proposed system incorporates a user-facing prediction and explanation layer.

---

## 🧠 Proposed Solution

The proposed framework combines:

- Data preprocessing
- Feature analysis
- Class imbalance handling
- Multiple machine learning models
- Ensemble learning
- Hyperparameter optimization
- Explainable AI
- Risk-oriented output
- User interface

### Proposed Pipeline

    CTG Dataset
         │
         ▼
    Data Preprocessing
         │
         ▼
    Feature Analysis
         │
         ▼
    Class Imbalance Handling
         │
         ▼
    Train / Validation Strategy
         │
         ▼
    Multiple ML Models
         │
         ▼
    Hyperparameter Optimization
         │
         ▼
    Ensemble Learning
         │
         ▼
    Fetal Health Prediction
         │
         ▼
    Explainable AI
         │
         ▼
    Risk Assessment
         │
         ▼
    User Interface
         │
         ▼
    Clinical Decision Support

---

## 📊 Dataset

The primary dataset considered for this project is the **Cardiotocography dataset from the UCI Machine Learning Repository**, as used in the base research.

### Dataset Characteristics

| Property | Description |
|---|---|
| Records | 2,126 |
| Input attributes | 21 |
| Target classes | 3 |
| Data type | Structured CTG-derived features |
| Main physiological information | Fetal Heart Rate and Uterine Contractions |

### Target Classes

The fetal health classification consists of:

| Class | Meaning |
|---|---|
| 0 | Normal |
| 1 | Suspect |
| 2 | Pathological |

### Class Distribution

The dataset contains approximately:

- **Normal:** 1,655 records
- **Suspect:** 295 records
- **Pathological:** 176 records

This imbalance is one of the important reasons for using imbalance-aware training and evaluation methods.

---

## 🧾 CTG Features

The dataset contains 21 input attributes representing different characteristics of fetal heart rate, fetal movement, uterine contractions and FHR histogram information.

Important features include:

- LB — Fetal Heart Beats per minute
- AC — Accelerations
- FM — Fetal Movements
- UC — Uterine Contractions
- LD — Light Decelerations
- SD — Severe Decelerations
- PD — Prolonged Decelerations
- ASTV — Percentage of abnormal short-term variability
- MSTV — Mean short-term variability
- ALTV — Percentage of abnormal long-term variability
- MLTV — Mean long-term variability
- HW — Histogram Width
- HM — Histogram Maximum
- HMin — Histogram Minimum
- NP — Histogram Peaks
- NZ — Histogram Zeroes
- HMo — Histogram Mode
- HMe — Histogram Mean
- HMed — Histogram Median
- HV — Histogram Variance
- HT — Histogram Tendency

The target variable is:

- NSP — Fetal state classification

---

## ⚙️ Data Preprocessing

The preprocessing stage may include:

1. Dataset loading
2. Data inspection
3. Checking for missing values
4. Duplicate inspection
5. Feature-target separation
6. Correlation analysis
7. Feature selection/reduction where required
8. Class distribution analysis
9. Train-test or cross-validation preparation

The base paper identified redundancy involving the Histogram Mean feature and removed it because of high correlation with other features.

The exact preprocessing and feature-selection strategy of the proposed system will be determined and documented based on experimental evaluation.

---

### SMOTE

The project considers **SMOTE — Synthetic Minority Oversampling Technique** for addressing class imbalance.

SMOTE generates synthetic minority-class examples rather than simply duplicating existing minority samples.

Conceptually:

    Minority Sample A
           +
    Minority Neighbour B
           │
           ▼
       SMOTE
           │
           ▼
    Synthetic Sample

SMOTE should be applied only to the training portion of the data.

The test data should remain untouched so that evaluation represents performance on unseen data.

---

## 🔄 Cross-Validation

The framework can use **5-fold cross-validation** to obtain a more robust estimate of model performance.

The dataset is divided into five folds.

    Fold 1 → Test
    Fold 2 + 3 + 4 + 5 → Train

    Fold 2 → Test
    Fold 1 + 3 + 4 + 5 → Train

    Fold 3 → Test
    Fold 1 + 2 + 4 + 5 → Train

    Fold 4 → Test
    Fold 1 + 2 + 3 + 5 → Train

    Fold 5 → Test
    Fold 1 + 2 + 3 + 4 → Train

Each fold is used once as the validation/test portion.

When SMOTE is used with cross-validation, oversampling should be performed only on the training portion of each fold to avoid data leakage.

---

## 🤖 Machine Learning Models

The proposed framework can evaluate multiple machine learning algorithms, including models investigated in the base study.

Potential candidate models include:

- Decision Tree
- Random Forest
- K-Nearest Neighbors
- Support Vector Machine
- Extra Trees
- Gradient Boosting
- LightGBM
- CatBoost

Different algorithms have different learning characteristics.

For example:

- Decision Trees learn hierarchical decision rules.
- Random Forest combines multiple decision trees.
- Extra Trees introduces additional randomization.
- SVM searches for separating decision boundaries.
- KNN makes predictions using neighboring samples.
- Gradient Boosting builds models sequentially to correct previous errors.
- LightGBM uses efficient gradient boosting.
- CatBoost is a gradient boosting algorithm designed to provide strong performance with structured data.

---

## 🧩 Ensemble Learning

A major component of the proposed framework is **ensemble learning**.

Instead of depending on a single model, multiple trained models can be combined to produce a final prediction.

Conceptually:

    Model 1 ─────┐
    Model 2 ─────┤
    Model 3 ─────┼──► Ensemble ───► Final Prediction
    Model 4 ─────┤
    Model 5 ─────┘

The purpose is to investigate whether combining complementary models can provide a more robust prediction than an individual classifier.

Possible ensemble strategies include:

- Voting
- Soft Voting
- Weighted Voting
- Stacking

The final ensemble strategy will be selected based on experimental evaluation.

---

## 🎯 Model Optimization

The proposed framework includes an optimization stage for identifying suitable model configurations.

Parameters that may be optimized include:

- Number of estimators
- Tree depth
- Learning rate
- Minimum samples per split
- Number of features
- Regularization parameters
- Kernel parameters
- Ensemble weights

Optimization techniques may include:

- Grid Search
- Random Search
- Bayesian Optimization
- Other suitable hyperparameter optimization methods

The optimization method will be selected according to computational requirements and experimental results.

---

## 📈 Evaluation Metrics

Because the dataset is imbalanced, the project will not rely only on conventional accuracy.

The following metrics can be considered:

### Accuracy

    Accuracy =
    Correct Predictions / Total Predictions

Accuracy measures the overall proportion of correct predictions.

However, on an imbalanced dataset, high accuracy may hide poor performance on minority classes.

### Recall

For a particular class:

    Recall = TP / (TP + FN)

Recall measures how many actual samples of a class were correctly identified.

### Balanced Accuracy

Balanced accuracy gives equal importance to each class.

For three classes:

    Balanced Accuracy =
    (Recall_Normal +
     Recall_Suspect +
     Recall_Pathological) / 3

This is particularly important for imbalanced fetal-health classification.

### Precision

    Precision = TP / (TP + FP)

Precision measures how many predicted positive samples are actually positive.

### F1-Score

    F1 = 2 × Precision × Recall /
         (Precision + Recall)

F1-score provides a balance between precision and recall.

### Confusion Matrix

The confusion matrix will be used to identify:

- Correct Normal predictions
- Normal classified as Suspect
- Normal classified as Pathological
- Suspect classified as Normal
- Suspect classified as Pathological
- Pathological classified as Normal
- Pathological classified as Suspect
- Correct Pathological predictions

---

## 🔍 Explainable AI

A key feature of the proposed system is **Explainable Artificial Intelligence (XAI)**.

The objective is not only to provide:

    Prediction: Pathological

but also to provide information about the factors that contributed to the prediction.

Potential XAI methods include:

- SHAP
- LIME
- Feature importance
- Local explanations
- Global feature analysis

### Example

    Prediction:
    Pathological

    Important contributing features:
    - Abnormal short-term variability
    - Abnormal long-term variability
    - FHR-related characteristics
    - Deceleration-related features

The actual explanations shown by the final system will be generated from the trained model and should not be manually assumed.

---

## 🩺 Risk Assessment

The system is intended to provide an AI-assisted assessment of fetal health condition/risk based on CTG-derived features.

The model output may be represented as:

    Normal
    Suspect
    Pathological

The interface may additionally present model confidence/probability values where appropriate.

The risk presentation should be treated as **decision support**, not as an independent clinical diagnosis.

---

## 🖥️ User Interface

The proposed system may provide a simple interface through which users can:

1. Enter or upload CTG-derived feature values.
2. Submit the input to the trained model.
3. Obtain a predicted fetal health class.
4. View prediction probabilities where available.
5. View important contributing features.
6. Interpret the model explanation.
7. Review the result in a simple dashboard.

Example workflow:

    CTG Feature Input
           ↓
    Preprocessing
           ↓
    Optimized Ensemble Model
           ↓
    Prediction
           ↓
    Probability / Risk Information
           ↓
    XAI Explanation
           ↓
    User Interface

---

## 🧪 Experimental Workflow

The overall experimental workflow is:

### Step 1 — Data Collection

Obtain the CTG dataset.

### Step 2 — Data Understanding

Analyze:

- Dataset dimensions
- Feature types
- Class distribution
- Missing values
- Correlations

### Step 3 — Preprocessing

Perform the required preprocessing operations.

### Step 4 — Data Splitting / Cross-Validation

Prepare training and evaluation data.

### Step 5 — Class Imbalance Handling

Apply SMOTE only to training data when required.

### Step 6 — Baseline Models

Train individual machine learning models.

### Step 7 — Hyperparameter Optimization

Optimize selected models.

### Step 8 — Ensemble Construction

Combine suitable models using an ensemble strategy.

### Step 9 — Evaluation

Evaluate models using:

- Accuracy
- Precision
- Recall
- F1-score
- Balanced Accuracy
- Confusion Matrix

### Step 10 — Explainability

Apply XAI methods to understand model predictions.

### Step 11 — Model Selection

Select the final model/framework based on predefined experimental criteria rather than assuming a particular algorithm will perform best.

### Step 12 — Interface Development

Integrate the selected model and explanation mechanism into a user interface.

---

### Programming Language

- Python

### Data Processing

- Pandas
- NumPy

### Machine Learning

- Scikit-learn
- LightGBM
- CatBoost

### Imbalanced Learning

- imbalanced-learn
- SMOTE

### Explainable AI

Potentially:

- SHAP
- LIME

### Visualization

Potentially:

- Matplotlib
- Seaborn

### Development Environment

Potentially:

- Jupyter Notebook
- Google Colab
- Visual Studio Code

### User Interface

Depending on the final implementation:

- Streamlit
- Flask
- FastAPI
- HTML/CSS/JavaScript

The final technology stack will depend on implementation requirements.

---

## 📁 Project Structure

```text
project-root/
│
├── data/
│   ├── raw/
│   └── processed/
│
├── notebooks/
│   ├── data_analysis.ipynb
│   ├── preprocessing.ipynb
│   ├── model_training.ipynb
│   └── explainability.ipynb
│
├── src/
│   ├── preprocessing.py
│   ├── feature_selection.py
│   ├── imbalance.py
│   ├── models.py
│   ├── optimization.py
│   ├── ensemble.py
│   ├── evaluation.py
│   └── explainability.py
│
├── models/
│   └── trained_models/
│
├── results/
│   ├── metrics/
│   ├── confusion_matrices/
│   └── plots/
│
├── app/
│   └── app.py
│
├── requirements.txt
│
├── README.md
│
└── LICENSE
